src/segmented_buffer.rs

```diff
@@ -737,6 +763,78 @@ mod test {
 
     use proptest::prelude::*;
 
+    #[cfg(test)]
+    proptest! {
+        #[test]
+        fn buf_impls (
+            inp in (
+                (0..100*1024usize), // data size
+                (0..8*1024usize), // segment size
+            )
+                .prop_flat_map(|(size, segment_size)|{
+                    (Just(size),
+                     Just(segment_size),
+                     (segment_size .. 10*segment_size),
+                     proptest::collection::vec(proptest::num::u8::ANY, size),
+                     )
+                })) {
+
+            let (size, segment_size, initial_capacity, values) = inp;
+
+            let mut buf = SegmentedPoolBufBuilder::new().segment_size(segment_size).initial_capacity(initial_capacity).build();
+
+            buf.write_all(&values).unwrap();
+            // Check that all the bytes were written to the buffer
+            assert_eq!(buf.len(), size);
+
+            // Check buf remaining for Buf impl
+            assert_eq!(buf.remaining(), size);
+
+            let buf_remaining = buf.remaining();
+
+            let mut reader = buf.buf.bytes_reader();
+            // Check same for reader struct Buf impl
+            assert_eq!(reader.remaining(), size);
+
+            let max_step = std::cmp::min(200, segment_size);
+
+            // walk over the reader checking that there is data available when there should be
+            let mut count = 0;
+            while count < size {
+                let step = std::cmp::min(max_step, reader.remaining());
+                assert!(reader.chunk().len() > 0);
+                reader.advance(step);
+                count += step;
+            }
```

Comment:
Shouldn't the logic be?

```rust
assert!(reader.remaining() > 0);

while reader.remaining() > 0 {
  let item = reader.chunk();
  assert!(item.len() > 0);
  reader.advance(item.len());
}

```

Reply:
That's the correct way to read it to not get duplicates, I'm deliberately walking across it in fixed size steps so that it would run into edge conditions when the segment size varies.

I can add an additional variant which walks across it as efficiently as possible

Reply:
I think we should verify that whats written is what's later read using the common read logic. Without fuzzying, the tests could be:
- "When I write a value across a single segment, the value is retrieved when read and internal state is what expected"
- "When I write a value across a multiple segments..."